src/Data/FunctorWithIndex.purs

```diff
@@ -0,0 +1,58 @@
+module Data.FunctorWithIndex
+  ( class FunctorWithIndex, imap
+  ) where
+
+import Prelude
+
+import Data.Maybe (Maybe)
+import Data.Maybe.First (First)
+import Data.Maybe.Last (Last)
+import Data.Monoid.Additive (Additive)
+import Data.Monoid.Conj (Conj)
+import Data.Monoid.Disj (Disj)
+import Data.Monoid.Dual (Dual)
+import Data.Monoid.Multiplicative (Multiplicative)
+
+
+-- | A `Functor` with an additional index.
+-- | Instances must satisfy a modified form of the `Functor` laws: 
+-- | 
+-- | - Identity: `imap (\_ a -> a) = id`
+-- | - Composition: `imap f . imap g = imap (\i -> f i <<< g i)`
+-- |
+-- | TODO: Don't we also want
+-- | - `imap (const f) = map f`
+-- | though?
+-- |
+-- | TODO: `imap` collides with `Invariant`s `imap`.
```

Comment:
Maybe we should either use `WithIndex` suffixes on all of these functions, or prefix with `ix` as an alternative to `i` prefixes?

I wouldn't be too worried about overlapping with `Invariant`'s map either though, there's always qualification. And `Invariant` is one of the less frequently used classes.